src/hotspot/share/cds/metaspaceShared.cpp

```diff
@@ -1367,6 +1367,8 @@ char* MetaspaceShared::reserve_address_space_for_archives(FileMapInfo* static_ma
     MemTracker::record_virtual_memory_split_reserved(total_space_rs.base(), total_space_rs.size(),
                                                      ccs_begin_offset, mtClassShared, mtClass);
   }
+  assert(archive_space_rs.is_reserved(), "Archive space is not reserved.");
```

Comment:
The logic is that, a large region `total_space_rs` is reserved and then is split into two sub regions. It doesn't matter what is the flag for `total_space_rs`. At split time the flags are set correctly for sub regions.

Reply:
The flags sent to the NMT subsystem is correct, but the flags recorded in the ReservedSpaces will be wrong, AFAIKT. You can probably verify that by adding asserts.

Reply:
If your comment refers only to these lines of code, they are already verified. Since, inside the split function, the sub-regions get the new flags and all the reserved and committed amounts are moved from the large region to the new ones. So, the accounting of memory is correct.

FWIW, if we trace down the call at line 1346 of `total_space_rs = Metaspace::reserve_address_space_for_compressed_classes(total_range_size, false /* optimize_for_zero_base */);` the region may get different flags of `mtClass` or `mtMetaspace` based on the checked criteria down there.
 
If you comment on all such cases, then I will double check for them and add assertion for.

Reply:
My point is that the `archive_space_rs` and `class_space_rs` can get the wrong flags assigned to them. The split functions don't change them. Right?

I would like to see the code run through our testing with these checks:
```
  assert(archive_space_rs.nmt_flag() == mtClassShared, "Sanity");
  assert(class_space_rs.nmt_flag() == mtClass, "Sanity");
```

Reply:
The call to `MemTracker::record_virtual_memory_split_reserved` at line 1364, takes two flags for the split parts. The corresponding regions in NMT take that flags.
The sanity assertions will be added anyway.

Reply:
I think you are still missing my point. `record_virtual_memory_split_reserved` doesn't update the `archive_space_rs` and `class_space_rs` instances with the correct flag.